File: 1.two-sum.java
/*
 * @lc app=leetcode id=1 lang=java
 *
 * [1] Two Sum
 */

// @lc code=start
import java.util.HashMap;
import java.util.Map;

class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        
    }
}

class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for ( int i = 0; i < nums.length; i++ ) 
        {
            map.put(nums[i], i);
        }
        for (int j = 0; j < nums.length; j++) 
        {
            int complement = target - nums[j];
            if ( map.containsKey( complement ) &&  map.get(complement) != j ) 
            {
                return new int[] { j, map.get( complement ) };
            }
        }

        return null;
    }
}
// @lc code=end

File: 1.two-sum.cpp
/*
 * @lc app=leetcode id=1 lang=cpp
 *
 * [1] Two Sum
 */

// @lc code=start
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> map = unordered_map<int, int>(nums.size());
        vector<int> res;

        for (int i = 0; i < nums.size(); i++) {
            if (map.count(nums[i])) {
                res.push_back(map[nums[i]]);
                res.push_back(i);
                break;
            }

            int key = target - nums[i];
            map[key] = i;
        }

        return res;
    }
};
// @lc code=end

